# Step 1 Data consistancy check

In [32]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [33]:
df_prods = pd.read_csv(r'C:\Users\jecka\2023-09-11 Achievement 4 Python\02 Data\Original Data\products.csv', index_col = False)

In [34]:
df_ords = pd.read_csv(r'C:\Users\jecka\2023-09-11 Achievement 4 Python\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [62]:
path=r'C:\Users\jecka\2023-09-11 Achievement 4 Python'

### Mixed Data

In [35]:
# Create a dataframe
df_test = pd.DataFrame()

In [36]:
# Create a Mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [37]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [38]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


### Missing Data

In [39]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [40]:
# finding products with missing names
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [41]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [42]:
# Checking row count in product dataframe
df_prods.shape

(49693, 5)

In [43]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [44]:
df_prods_clean.shape

(49677, 5)

### Duplications

In [45]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [46]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [47]:
# Addressing the Duplications
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
# I skipped the shape command as it can be seen above

In [48]:
df_prods_clean_no_dups.shape

(49672, 5)

# Step 2

In [49]:
# Run the df.describe() function on your df_ords dataframe
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

In the days_since_last_order column, the numbers are odd. The minimum is 0 which could be correct, but the first Quartile is 4, while the median is 7, the third Quartile is 1.5, and the max is 3. The Median and Maxium are the two that are the most interesting as the median cannot be higher than the maximum.

This is also true for the order_number column. This column is for the total orders but the first Quartile is 5, while the median is 1.02, and the max is 2.06.

# Step 3

In [50]:
# checking for mixed data types 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# Step 4
##### There are no values with mixed data types.

# Step 5

In [51]:
# Checking for missing data
df_ords.isnull().sum()

order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [52]:
# Finding the rows with missing calues
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [53]:
df_ords_nan

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329        1    prior             1                   2   
11        2168274        2    prior             1                   2   
26        1374495        3    prior             1                   1   
39        3343014        4    prior             1                   6   
45        2717275        5    prior             1                   3   
...           ...      ...      ...           ...                 ...   
3420930    969311   206205    prior             1                   4   
3420934   3189322   206206    prior             1                   3   
3421002   2166133   206207    prior             1                   6   
3421019   2227043   206208    prior             1                   1   
3421069   3154581   206209    prior             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

In [54]:
df_ords_nan.describe()

order_id        user_id  order_number  orders_day_of_week  \
count  2.062090e+05  206209.000000      206209.0       206209.000000   
mean   1.708462e+06  103105.000000           1.0            2.754118   
std    9.881299e+05   59527.555167           0.0            2.076205   
min    2.000000e+01       1.000000           1.0            0.000000   
25%    8.507300e+05   51553.000000           1.0            1.000000   
50%    1.706246e+06  103105.000000           1.0            3.000000   
75%    2.564292e+06  154657.000000           1.0            5.000000   
max    3.421081e+06  206209.000000           1.0            6.000000   

       order_hour_of_day  days_since_last_order  
count      206209.000000                    0.0  
mean           13.626597                    NaN  
std             4.223769                    NaN  
min             0.000000                    NaN  
25%            11.000000                    NaN  
50%            14.000000                    NaN  
75%            17.000000                    NaN  
max            23.000000                    NaN

The missing values are most likely the cause of first time and one time customers. Only one order was ever placed so the df_ords_nan column would remain blank as there would not be any prior orders or time between orders.

# Step 6 Addressing Missing values
##### I decided to leave the missing values in. Although it's a fairly large amount it's not enough to filter out and the rows are important as they are are for customers who only placed one order.

# Step 7

In [55]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [56]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

There are no duplicates in the orders data frame.

# Step 8
##### There are no duplicates so nothing was done. However if there were I would filter them out and move the non duplicated rows into a new data frame.

# Step 9 Export files

In [59]:
# Changing the name because I didn't do it earlier
df_prods_checked = df_prods_clean_no_dups

In [67]:
# Export product.csv
df_prods_checked.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'), index=False)

In [68]:
# Export orders.csv
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index=False)